OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs

```diff
@@ -57,25 +58,39 @@ public override bool Produce(Actor self, ActorInfo producee, string productionTy
 			CPos endPos;
 			WAngle spawnFacing;
 
+			var bounds = map.Bounds;
+			var diagonal = Exts.ISqrt(bounds.Height * bounds.Height + bounds.Width * bounds.Width);
+
 			if (info.BaselineSpawn && mpStart != null)
 			{
 				var spawn = mpStart.Start[owner];
-				var bounds = map.Bounds;
 				var center = new MPos(bounds.Left + bounds.Width / 2, bounds.Top + bounds.Height / 2).ToCPos(map);
 				var spawnVec = spawn - center;
-				startPos = spawn + spawnVec * (Exts.ISqrt((bounds.Height * bounds.Height + bounds.Width * bounds.Width) / (4 * spawnVec.LengthSquared)));
+				startPos = spawn + spawnVec * diagonal / (2 * spawnVec.Length);
 				endPos = startPos;
-				var spawnDirection = new WVec((self.Location - startPos).X, (self.Location - startPos).Y, 0);
-				spawnFacing = spawnDirection.Yaw;
+				spawnFacing = map.FacingBetween(startPos, self.Location, WAngle.Zero);
 			}
 			else
 			{
-				// Start a fixed distance away: the width of the map.
-				// This makes the production timing independent of spawnpoint
-				var loc = self.Location.ToMPos(map);
-				startPos = new MPos(loc.U + map.Bounds.Width, loc.V).ToCPos(map);
-				endPos = new MPos(map.Bounds.Left, loc.V).ToCPos(map);
+				// Start a fixed distance away: the longest straight line across the
+				// map at a given angle. This makes the production timing independent
+				// of spawnpoint.
 				spawnFacing = info.Facing;
+				var diagonalAngle = WAngle.ArcTan(bounds.Width, bounds.Height);
+				var angle = Math.Abs(spawnFacing.Angle2);
+
+				if (angle > 256)
+					angle = 512 - angle;
+
+				int longestApproach;
+				if (angle < Math.Abs(diagonalAngle.Angle2))
+					longestApproach = bounds.Height * 1024 / spawnFacing.Cos();
+				else
+					longestApproach = bounds.Width * 1024 / spawnFacing.Sin();
+
+				var spawnVec = new MPos(spawnFacing.Sin() * longestApproach / 1024, spawnFacing.Cos() * longestApproach / 1024).ToCPos(map) - CPos.Zero;
```

Comment:
`MPos` is an abstract coordinate - creating them based on world distances/angles is bogus!

Reply:
That is not new in this PR! The trait has always done geometry with grid coordinates. This is hacky, but I just use what I inherited. I've pointed out before that this is wrong and we should just use `WPos` throughout the entire trait, just like `AirstrikePower` already does. This is something I already planned on doing, but I don't see why that should derail this PR.

Reply:
The previous code was equating `PPos` and `MPos` which is still bogus but mostly harmless because they follow the same geometrical conventions. World coordinates and MPos/PPos don't!

There shouldn't be much difficulty in doing this fully in world coordinates:
1. Use `Map.ProjectedTopLeft` and `Map.ProjectedBottomRight` to calculate the width and height of the map in world coordinates
2. Calculate the hypotenuse length for triangles with the horizontal/vertical edge matching the world width/height
3. The *minimum* (not maximum as I said above, temporarily confused by the `longestApproach` name) of these two numbers defines the length of the flight vector.
4. `new WVec(0, -length, 0).Rotate(spawnFacing);` should gives the vector offset (or something which is wrong only in sign/axis) to add to the destination.